torchrl/objectives/crossq.py

```diff
@@ -0,0 +1,691 @@
+        td_q = self._vmap_qnetworkN0(
+            td_q,
+            self._cached_detached_qvalue_params,  # should we clone?
+        )
+        min_q_logprob = (
+            td_q.get(self.tensor_keys.state_action_value).min(0)[0].squeeze(-1)
+        )
+
+        if log_prob.shape != min_q_logprob.shape:
+            raise RuntimeError(
+                f"Losses shape mismatch: {log_prob.shape} and {min_q_logprob.shape}"
+            )
+
+        return self._alpha * log_prob - min_q_logprob, {"log_prob": log_prob.detach()}
+
+    def _compute_target(self, tensordict) -> Tensor:
+        r"""Value network for CrossQ.
+
+        CrossQ is based on a value estimate of the form:
+
+        .. math::
+
+          V = Q(s,a) - \alpha * \log p(a | s)
+
+        This class computes this value given the actor and qvalue network
+
+        """
+        tensordict = tensordict.clone(False)
+        # get actions and log-probs
+        with torch.no_grad():
+            with set_exploration_type(
+                ExplorationType.RANDOM
+            ), self.actor_network_params.to_module(self.actor_network):
+                next_tensordict = tensordict.get("next").clone(False)
+                next_dist = self.actor_network.get_dist(next_tensordict)
+                next_action = next_dist.rsample()
+                next_tensordict.set(self.tensor_keys.action, next_action)
+                next_sample_log_prob = next_dist.log_prob(next_action)
+
+        # get q-values
+        next_tensordict_expand = self._vmap_qnetworkN0(
+            next_tensordict, self.qvalue_network_params
+        )
+        state_action_value = next_tensordict_expand.get(
+            self.tensor_keys.state_action_value
+        )
+        if (
+            state_action_value.shape[-len(next_sample_log_prob.shape) :]
+            != next_sample_log_prob.shape
+        ):
+            next_sample_log_prob = next_sample_log_prob.unsqueeze(-1)
+        next_state_value = state_action_value - self._alpha * next_sample_log_prob
+        next_state_value = next_state_value.min(0)[0]
+        tensordict.set(
+            ("next", self.value_estimator.tensor_keys.value), next_state_value
+        )
+        target_value = self.value_estimator.value_estimate(tensordict).squeeze(-1)
+        return target_value.detach()
+
+    def _qvalue_loss(
+        self, tensordict: TensorDictBase
+    ) -> Tuple[Tensor, Dict[str, Tensor]]:
+        # we pass the alpha value to the tensordict. Since it's a scalar, we must erase the batch-size first.
+
+        target_value = self._compute_target(tensordict)
+        tensordict_expand = self._vmap_qnetworkN0(
+            tensordict.select(*self.qvalue_network.in_keys, strict=False),
+            self.qvalue_network_params,
+        )
+        pred_val = tensordict_expand.get(self.tensor_keys.state_action_value).squeeze(
+            -1
+        )
+
+        # ############################
+        # # compute next action
```

Comment:
This is quite confusing to me. When computing the state_action_values and next_state_action_values combined as shown below CrossQ does not converge and is slower compared to the separate target_value computation. I checked the values for the pred_val and target_values with next_action = dist.loc for both versions and they are the same, as expected. Would be great if you could have a look @vmoens